src/solvers/petsc_nonlinear_solver.C

```diff
@@ -825,6 +876,38 @@ PetscNonlinearSolver<T>::solve (SparseMatrix<T> &  pre_in,  // System Preconditi
   // Should actually be a PetscReal, but I don't know which version of PETSc first introduced PetscReal
   Real final_residual_norm=0.;
 
+  // We don't want to do this twice because it resets
+  // SNESSetLagPreconditioner
+  if ((this->reuse_preconditioner) && (!_setup_reuse))
+    {
+      _setup_reuse = true;
+      ierr = SNESSetLagPreconditionerPersists(_snes, PETSC_TRUE);
+      LIBMESH_CHKERR(ierr);
+      // According to the PETSC 3.16.5 docs -2 is a magic number which
+      // means "recalculate the next time you need it and then not again"
+      ierr = SNESSetLagPreconditioner(_snes, -2);
+      LIBMESH_CHKERR(ierr);
+      // Add in our callback which will trigger recalculating
+      // the preconditioner when we hit reuse_preconditioner_max_its
+      ierr = SNESMonitorSet(_snes, &libmesh_petsc_recalculate_monitor,
+                            (void*)
+                            &(this->reuse_preconditioner_max_its),
+                            NULL);
+      LIBMESH_CHKERR(ierr);
+    }
+  else if (!(this->reuse_preconditioner))
+    // This covers the case where it was enabled but was then disabled
+    {
+      ierr = SNESSetLagPreconditionerPersists(_snes, PETSC_FALSE);
+      LIBMESH_CHKERR(ierr);
+      if (_setup_reuse)
+        {
+          _setup_reuse = false;
+          ierr = SNESMonitorCancel(_snes);
```

Comment:
We can not really do this. This will kill all monitors

Reply:
Currently everything is re-setup each time, so the only monitor that might be canceled is the default monitor.  We should be able to fix this by adding

```
      if (_default_monitor)
        {
          ierr = SNESMonitorSet (_snes, libmesh_petsc_snes_monitor,
                                 this, PETSC_NULL);
          LIBMESH_CHKERR(ierr);
        }
```

after the this, right?  Or better move that to the `make_and_add_default_monitor` method.